gk/main.c

```diff
@@ -52,25 +351,117 @@ gk_proc(void *arg)
 		 */
 
 		/* Get burst of RX packets, from first port of pair. */
+		int i;
+		int ret = -1;
 		uint16_t num_rx;
-		uint16_t num_tx;
-		struct rte_mbuf *bufs[GATEKEEPER_MAX_PKT_BURST];
+		uint16_t num_tx = 0;
+		uint16_t num_tx_succ;
+		struct rte_mbuf *rx_bufs[GATEKEEPER_MAX_PKT_BURST];
+		struct rte_mbuf *tx_bufs[GATEKEEPER_MAX_PKT_BURST];
 
-		num_rx = rte_eth_rx_burst(port_in, 0, bufs,
+		/* Load a set of packets from the front NIC. */
+		num_rx = rte_eth_rx_burst(port_in, 0, rx_bufs,
 			GATEKEEPER_MAX_PKT_BURST);
 
 		if (unlikely(num_rx == 0))
 			continue;
 
+		/* For each packet in this set do the following. */
+		for (i = 0; i < num_rx; i++) {
+			struct ipacket packet;
+			/*
+			 * Pointer to the flow entry in request state 
+			 * under evaluation.
+			 */
+			struct flow_entry *fe;
+			struct rte_mbuf *pkt = rx_bufs[i];
+			
+			extract_packet_info(pkt, &packet);
+
+			/* 
+			 * Find the flow entry for the IP pair.
+			 * Create a new flow entry if not found.
+			 */
+			ret = rte_hash_lookup(instance->ip_flow_hash_table,
+				&packet.flow);
+			if (ret < 0) {
+				/* Create a new flow entry. */
+				ret = rte_hash_add_key(instance->ip_flow_hash_table,
+ 					(void *)&packet.flow);
+				if (ret < 0) {
+					RTE_LOG(ERR, HASH,
+						"The GK block failed to add new key to hash table!\n");
+					continue;
+				}
+
+				fe = &instance->ip_flow_entry_table[ret];
+				initialize_flow_entry(fe, &packet.flow);
+			} else
+				fe = &instance->ip_flow_entry_table[ret];
+
+			/*
+			 * 1.1 If the pair of source and destination addresses 
+			 * is in the flow table, proceed as the entry instructs,
+			 * and go to the next packet.
+			 */
+			switch(fe->state) {
+				case GK_REQUEST:
```

Comment:
All of these `case` statements (and the bodies of code within them) can be pulled back one level of indentation.